### PR TITLE
📦 onnx support for santacoder

### PR DESCRIPTION
**Pull request by mayank31398**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,11 +1,11 @@
 repos:
   - repo: https://github.com/pycqa/isort
-    rev: 5.10.1
+    rev: 5.12.0
```
Comment (jlamypoirier) [inline .pre-commit-config.yaml:3]: This doesn't match requirement.txt

File: Makefile [hunk 2]
```diff
@@ -5,16 +5,16 @@ style:
 	black --preview $(check_dirs)
 	isort $(check_dirs)
 
-BATCH_SIZE ?= 1
-DTYPE ?= float16
+BATCH_SIZE ?= 4
```
Comment (jlamypoirier) [inline Makefile]: Please don't change the defaults

File: Makefile [hunk 3]
```diff
@@ -5,16 +5,16 @@ style:
...
 HIDDEN_SIZE ?= 2048
 N_HEAD ?= 16
 N_LAYER ?= 24
 N_POSITION ?= 2048
-MAX_INPUT_LENGTH ?= -1
+MAX_INPUT_LENGTH ?= 512
 
-RUN_HF := python3 src/main.py --pipeline_class=HF_Pipeline
-RUN_DS := deepspeed --num_gpus 1 src/main.py --pipeline_class=DS_Pipeline
+RUN_HF := python -m src.main --pipeline_class=HF_Pipeline
```
Comment (jlamypoirier) [inline Makefile]: It's not an installed package so we should refer to it by path, not module

File: Makefile [hunk 4]
```diff
@@ -66,3 +66,19 @@ santacoder:
 .PHONY: optimized-santacoder
 optimized-santacoder:
 	${RUN_HF} --pretrained_model=olivierdehaene/optimized-santacoder --tokenizer=bigcode/santacoder --trust_remote_code ${EXP_ARGS}
+
+.PHONY: santacoder-cpu
```
Comment (jlamypoirier) [inline Makefile]: No need to have different targets for cpu and gpu, $EXTRA_ARGS already covers it. 
Why do we need `mayank31398/santacoder`? How is it different from the other targets above?
Please build the commands like the other ones above.

File: src/pipeline.py [hunk 5]
```diff
@@ -266,9 +267,61 @@ def __init__(self, **kwargs):
         )
 
 
+class ONNX_Pipeline(Pipeline):
+    def __init__(
```
Comment (jlamypoirier) [inline src/pipeline.py:271]: The vast majority of this function is identical to the base class. Seems like you just need to override methods like `_get_config, `_create_model`, etc. and add some assertions like in `DS_Pipeline`

File: src/pipeline.py [hunk 6]
```diff
@@ -266,9 +267,61 @@ def __init__(self, **kwargs):
...
+        log_rank_n("*** Setting up tokenizer", logger.info)
+        t0 = time.perf_counter()
+        self.tokenizer = AutoTokenizer.from_pretrained(tokenizer)
+
+        self.tokenizer.add_special_tokens({"pad_token": "[PAD]"})
+        t1 = time.perf_counter()
+
+        self.device = device
+
+        if device == torch.device("cpu"):
```
Comment (jlamypoirier) [inline src/pipeline.py:294]: Missing `else`

File: src/pipeline.py [hunk 7]
```diff
@@ -266,9 +267,61 @@ def __init__(self, **kwargs):
...
+            provider = "CPUExecutionProvider"
+        elif device == torch.device("cuda"):
+            provider = "CUDAExecutionProvider"
+
+        self.dtype = dtype
+
+        if self.dtype != torch.float32:
```
Comment (jlamypoirier) [inline src/pipeline.py:303]: It still doesn't support float16?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mayank31398) [reply .pre-commit-config.yaml:3]: addressed

File: Makefile (hunk 2 above)
Comment (mayank31398) [reply Makefile]: yeah, I was experimenting, sorry will revert this one

File: src/pipeline.py (hunk 7 above)
Comment (mayank31398) [reply src/pipeline.py:303]: it does but not for santacoder. it requires a merged graph (at least if you compile via optimum)

File: Makefile (hunk 3 above)
Comment (mayank31398) [reply Makefile]:
```python
python3 src/main.py --pipeline_class=ONNX_Pipeline --pretrained_model=mayank31398/santacoder --tokenizer=mayank31398/santacoder --trust_remote_code --dtype=float16 --batch_size=1 --max_input_length=-1 device=cuda
Traceback (most recent call last):
  File "src/main.py", line 11, in <module>
    from src.metrics import Metrics
ModuleNotFoundError: No module named 'src'
make: *** [Makefile:73: santacoder-onnx] Error 1
```
For me it doesn't work :)
are you pip installing it?

Comment (mayank31398) [reply Makefile]: Ill need to convert everything in the code to use relative paths to make this work I guess?

File: Makefile (hunk 4 above)
Comment (mayank31398) [reply Makefile]: `mayank/santacoder` has a fix for a bug. This bug is not encountered when running the vanilla generate method.
But for compiling the graph, onnx runs the model by passing dummy tensors and doesn't take the generate() path.
Its only a single line change in `modeling_gpt`

Comment (mayank31398) [reply Makefile]: I had created a PR for this: https://huggingface.co/bigcode/santacoder/discussions/23#640f55462ec22527ea436d5e

File: Makefile (hunk 3 above)
Comment (jlamypoirier) [reply Makefile]: Hmm I had another look, I guess you're right. The `-m` is needed to do the imports properly (though it still needs to be called from the repo root), in the dockerfile I'm just adding to the pythonpath. Relative paths wouldn't work either.

Please keep `python3` as the executable, `python` tends to refer to python2

**Pull request closed without merging**